syntax = "proto3";
package books;

// gRPC service for a replicated book-stock database
service BooksDatabase {
  rpc Read  (ReadRequest)  returns (ReadResponse);
  rpc Write (WriteRequest) returns (WriteResponse);
  rpc DecrementStock(DecrementRequest) returns (WriteResponse);
}


message DecrementRequest {
  string title = 1;
  int32  amount = 2;
}

message ReadRequest {
  string title = 1;
}
message ReadResponse {
  int32 stock = 1;
}

message WriteRequest {
  string title     = 1;
  int32  new_stock = 2;
}
message WriteResponse {
  bool success = 1;
}